Rename EffectGlobal to LightingGlobal and expose individual channels.

phrohdoh opened the pull request:
~~Without the `Lighting` prefix the property names are too generic in my opinion.~~

A new take on #8853. My effort was not to do this yet but rather to expose anything on the world.
Guess that's not the optimal way.

I can spin up a test-case in RA2 tomorrow if desired.

Mailaender commented:
Seems rather redundant.

pchote commented:
This fixes a limitation from #8684 where you can't query the current lighting state.

Looks sensible, but needs a testcase.

Mailaender commented:
I guess you can get rid of ChangeLighting then.

phrohdoh committed:
Rename EffectsGlobal to LightingGlobal
Remove ChangeLighting in favor of individual channels
Change RA's fort-lonestar to use Lighting instead of Effect

Mailaender commented on OpenRA.Mods.Common/Scripting/Global/EffectGlobal.cs:
```diff
@@ -1,51 +0,0 @@
-		}
-
-		[Desc("Controls the `FlashPaletteEffect` trait.")]
-		public void Flash(string type = null, int ticks = -1)
```
This unnecessary refactoring will break https://github.com/OpenRA/OpenRA/blob/bleed/mods/ra/maps/fort-lonestar/fort-lonestar.lua#L153.

Phrohdoh replied:
Sorry I had not caught that `Flash`. There is no need to have two tables that do the same thing.

Phrohdoh commented:
Fixed.

pchote commented on OpenRA.Mods.Common/Scripting/Global/LightingGlobal.cs:
```diff
@@ -0,0 +1,66 @@
+			lighting = context.World.WorldActor.TraitOrDefault<GlobalLightingPaletteEffect>();
+
+			if (lighting == null)
+				throw new LuaException("World actor must have GlobalLightingPaletteEffect to use the Lighting table.");
```
Global tables are always loaded, so this breaks all scripted maps!
RA/TD/D2K now all crash when trying to load their shellmaps.

pchote replied:
You can either return 0 or throw in the individual properties instead.

Phrohdoh replied:
Hmm checking for null (or a bool) in each property access is a bit distasteful, but okay.

pchote replied:
Less distasteful than completely breaking our RA and TD mods!

Phrohdoh replied:
I didn't say I wouldn't do it or ever compare it to that.

pchote commented:
This changes a so-far-unreleased Lua API, so we need to take this before the playtest.

Phrohdoh commented:
Fixed the breakage of TD and RA.

obrakmann commented:
Looks good to me :+1:

pchote commented:
:+1: after removing the test case.

Phrohdoh commented:
Testcase removed.

pchote commented:
https://github.com/OpenRA/OpenRA/wiki/Changelog-%28bleed%29/_compare/2870d06a00486575aab336a89980fe25445a3306%5E...2870d06a00486575aab336a89980fe25445a3306

Mailaender commented on OpenRA.Mods.Common/Scripting/Global/LightingGlobal.cs:
```diff
@@ -0,0 +1,64 @@
+
+		public double Red
+		{
+			get { return hasLighting ? lighting.Red : 0d; }
```
Shouldn't this be `1d` for unchanged lighting. I think 0 would be black.

Phrohdoh replied:
My idea was to indicate that this wasn't usable (getting a bogus value back).
But perhaps 1d would have been a better default.

Mailaender replied:
`null`/`nil` would be a quite unfriendly and `0d` a bit confusing. ;)
